### PR TITLE
Quiz support on Portal

### LINKED ISSUE #51: Quiz support on Portal
Supporting Quiz as a destination on Portal
- Includes adding an API key for every organization. 
- Quiz checks the presence of a user ID and an API key and validates the API key.

### PR DESCRIPTION
**Pull request by sudheshna-donthineni**
Fixes #51 

## Summary

This is the first PR in adding support for Quiz on Portal. 
This code adds two query parameters to a route: user ID and API key. 
It also checks to make sure they exist and are not null or undefined. 

## Test Plan

- [x] Test Responsiveness
  - [x] Laptop (1200px)
  - [x] Tablet (760px)
  - [x] Phone (320px)
- [ ] Cross-Browser Testing
  - [x] Chrome
  - [ ] Firefox
- [ ] ~Local Language Support~
- [x] Hygiene and Housekeeping
  - [x] Self-review
  - [x] Comments have been added appropriately
  - [ ] Check for bundle size [here](https://bundlephobia.com/) if adding a package
  - [x] Added relevant details like Labels/Projects/Milestones etc.
  - [x] If adding or removing any environment variable, update `docs/ENV.md`, `.env.example` and the Github workflows.
- [ ] Testing
  - [ ] Wrote tests
  - [x] Tested locally
  - [x] Tested on staging
  - [x] Tested on an actual physical phone
  - [ ] Tested on production
- [ ] ~Lighthouse Checks~
  - [ ] Images have `alt` attributes
  - [ ] Any `` tags have `width` and `height` specified
  - [ ] Any `target="_blank"` links have `rel="noopener"`
  - [ ] Only SVGs are used as images. If PNGs are used, their size has been optimised.
  - [ ] Any SVG buttons without text have their `aria-label` attributes set

Linked backend PR: https://github.com/avantifellows/quiz-backend/pull/42

For testing:
- Have created a dummy session.
http://localhost:8080/?sessionId=test-quiz
https://staging-auth.avantifellows.org/?sessionId=test-quiz

### CODE REVIEW

File: src/router/index.ts
Comment (deepansh96) [inline src/router/index.ts:10]: Instead of explicitly specifying `quizId` ka param, you can just do this
```js
 props: (route: any) => ({
      ...route.params,
      thirdPartyUserId: route.query.userId,
      thirdPartyApiKey: route.query.apiKey,
    }),
```

File: src/router/index.ts
Comment (deepansh96) [inline src/router/index.ts:12]: why not just `userId` and `apiKey`?
All our authentication in quiz will be third party only.

This naming made sense in plio because it had it's own authentication system and a third party one as well

File: src/services/API/User.ts
Comment (deepansh96) [inline src/services/API/User.ts:9]: how does this sound?
"authenticates the user using the provided API key"?

File: src/services/API/User.ts
Comment (deepansh96) [inline src/services/API/User.ts:11]:
```suggestion
   * @returns {Promise<UserAPIResponse>} org details corresponding to the user that is trying to log in
```

**Review by deepansh96 — CHANGES_REQUESTED** [review]

**Review by pritamps — CHANGES_REQUESTED** [review]

File: src/services/API/User.ts
Comment (pritamps) [inline src/services/API/User.ts:11]: Remove user from this comment. The quiz doesn't know anything about the link between the user and the org

File: src/services/API/User.ts
Comment (pritamps) [inline src/services/API/User.ts:4]: It shouldn't be called `UserAPIResponse`. Like I said above, the quiz has only info about whether an org key is valid or not. So this API has to do only with the org

**Review by pritamps — CHANGES_REQUESTED** [review]

File: src/router/index.ts
Comment (dalmia) [inline src/router/index.ts:58]: What we are really checking is whether the user is authenticated right?

```suggestion
    const isAuthenticated =
      requiredAuthKeys.every((key) => queryParams.includes(key) &&  (queryParams[key] != "" && queryParams[key] != undefined))
```

File: src/services/API/Endpoints.ts
Comment (dalmia) [inline src/services/API/Endpoints.ts:4]: why not simply?

+ if we are changing the endpoint in the backend based on my comments, we'll have to change that here too.

```suggestion
export const authenticationEndpoint = "/check-auth-token/"
```

File: src/router/index.ts
Comment (dalmia) [reply src/router/index.ts:51]: `requiresAuth` is a navigation guard that is placed on each route. So, here, we have placed `requiresAuth = true` for the quiz route (`/quiz`). So, one cannot access any quiz without the auth. But this allows the flexibility to enable someone to view some other page in the future where we don't need authentication instead of hard-coding the required authentication for every route.

File: src/types.ts
Comment (dalmia) [inline src/types.ts:157]: based on my comments in the backend

```suggestion
  name: string;
```

File: src/services/API/User.ts
Comment (dalmia) [inline src/services/API/User.ts:9]:
```suggestion
   * authenticates the user using the API key
```

File: src/router/index.ts
Comment (dalmia) [inline src/router/index.ts:13]: why do you need this? We have been reading userId as a query param directly in `Player.vue` so far. Why can't we continue doing that?

File: src/views/Player.vue
Comment (dalmia) [inline src/views/Player.vue:96]: same comment as above - why do we need them as prop - can directly read the query param as we've been reading userId before

File: src/views/Player.vue
Comment (dalmia) [inline src/views/Player.vue:143]: please check your linter settings - we should all have the same linter settings or else these non-changes will keep showing up as changes and make reviewing a pain. Please check with me or Deepansh for our linter settings.

File: src/views/Player.vue
Comment (dalmia) [inline src/views/Player.vue:122]: this is already done in `router/index.ts` no? There is no reason to repeat it here.

File: src/views/Player.vue
Comment (dalmia) [inline src/views/Player.vue:133]: you don't need to define a new function - you can directly write:
```suggestion
    UserAPIService.checkAuthToken(props.apiKey).catch(() => {
          router.replace({
            name: "403",
          });
        }); 
```

**Review by dalmia — CHANGES_REQUESTED** [review]

File: src/router/index.ts
Comment (sudheshna-donthineni) [reply src/router/index.ts:13]: It was being used once before. Now that's it being used in multiple places, I think having a variable is better, no?

**Review by pritamps — APPROVED** [review]
I'm okay with this -- Please get signoff from @dalmia / @deepansh96 also before merging

File: src/services/API/Organization.ts
Comment (deepansh96) [inline src/services/API/Organization.ts:4]:
```suggestion
  OrganizationAPIResponse
```

File: src/views/Player.vue
Comment (deepansh96) [inline src/views/Player.vue:68]:
```suggestion
import OrganizationAPIService from "../services/API/Organization";
```

File: src/router/index.ts
Comment (deepansh96) [reply src/router/index.ts:13]: It's fine by me. Will maybe wait for aman to respond

File: src/types.ts
Comment (deepansh96) [reply src/types.ts]: No no I can see
I thought you added it / changed it by mistake because this is not related to your feature right

File: tests/e2e/specs/renders-player-assessment.js
Comment (dalmia) [inline tests/e2e/specs/renders-player-assessment.js:30]: you don't need to give a valid api key - the check is happening on the backend right? you just need to give some value for the tests.

File: cypress.env.json
Comment (dalmia) [inline cypress.env.json:3]: this is also not needed because of my last comment

File: tests/e2e/specs/renders-player-assessment.js
Comment (dalmia) [inline tests/e2e/specs/renders-player-assessment.js:26]:
```suggestion
          "/organizations/authenticate/*",
        {
```

File: tests/e2e/specs/renders-player-assessment.js
Comment (dalmia) [inline tests/e2e/specs/renders-player-assessment.js:248]:
```suggestion
          "/organizations/authenticate/*",
```

File: tests/e2e/specs/renders-player-assessment.js
Comment (dalmia) [inline tests/e2e/specs/renders-player-assessment.js:253]: same comment as before

**Review by dalmia — APPROVED** [review]

**Pull request merged**
